### PR TITLE
history-view

### LINKED ISSUE #32: Feature Request: History panel
Having a history panel where you could go back to old equations or copy the expression/result would be super handy. An idea for implementation could be like the this example (http://www.eeweb.com/toolbox/calculator)

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/29684942-feature-request-history-panel?utm_campaign=plugin&utm_content=tracker%2F2163668&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F2163668&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### LINKED ISSUE #42: Multiline screen display
#### Expected behaviour
Viewing old results of earlier calculation with scrolling multi line. Old results are Non-editable.

### LINKED ISSUE #131: History View
#### Expected behaviour
The History View should locate above the entry box and be designed to show previous calculations in the form of a scrollable list that can be clicked and scrolled through using a mouse or even navigated and selected using the arrow keys on a keyboard.
Once an entry has been selected, it will immediately show up in the entry box for to manipulate and use.
Finally, there is no upper limit on the number of calculations that can show up in the list for use.

#### Actual behaviour
We can only access history in the MATE Calculator using the kayboard shortcut Undo command (**Ctrl**+**Z**)

#### Steps to reproduce the behaviour
Open MATE Calculator and do some calculations.

#### Package version
mate-calc 1.20.3

#### Linux Distribution
Debian 10.1

### PR DESCRIPTION
**Pull request by mbkma**
So this lets mate-calc as it is. It just adds the option to show the history if desired by the user (menu->View->History or ctrl+h).
- adds history view to show recent calculations.
- adds an option to enable/disable history (disabled by default).
- adds the possibility to resize mate-calc if history is enabled.

Closes #32 and closes #42 and closes #131

### CODE REVIEW

**Comment by sc0w** [comment]
you can squash both commits into one

File: src/math-equation.c [hunk 1]
```diff
@@ -1788,8 +1793,19 @@ math_equation_class_init(MathEquationClass *klass)
...
+			G_TYPE_FROM_CLASS(klass),
+			G_SIGNAL_RUN_LAST,
+			0,
+			NULL,
+			NULL,
+			NULL,
+			G_TYPE_NONE,
+			3,
+			param_types);
+}
```
Comment (sc0w) [inline src/math-equation.c]: wrong indentation

File: src/math-window.c [hunk 2]
```diff
@@ -549,6 +627,14 @@ math_window_class_init(MathWindowClass *klass)
...
                                                         G_PARAM_READWRITE | G_PARAM_CONSTRUCT_ONLY));
 
+    g_object_class_install_property(object_class,
+                                    PROP_EQUATION,
+                                    g_param_spec_boolean("show-history",
+                                                        "show-history",
+                                                        "Show-history",
+                                                        FALSE,
+                                                        G_PARAM_READWRITE));
+
```
Comment (sc0w) [inline src/math-window.c]: wrong indentation

File: src/history-entry.ui [hunk 3]
```diff
@@ -0,0 +1,83 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!-- Generated with glade 3.22.1 -->
+<interface>
+  <requires lib="gtk+" version="3.20"/>
```
Comment (sc0w) [inline src/history-entry.ui]: mate-calc requires gtk 3.22

**Comment by mbkma** [comment]
@sc0w done :)

**Comment by sc0w** [comment]
when I click twice in one entry at the history, I see in terminal:

```
(mate-calc:310019): Gtk-WARNING **: 01:14:52.807: Invalid text buffer iterator: either the iterator is uninitialized, or the characters/pixbufs/widgets in the buffer have been modified since the iterator was created.
You must use marks, character numbers, or line numbers to preserve a position across buffer modifications.
You can apply tags and insert marks without invalidating your iterators,
but any mutation that affects 'indexable' buffer contents (contents that can be referred to by character offset)
will invalidate all outstanding iterators
```

please make sure there is no warnings in terminal

**Comment by mbkma** [comment]
I guess you have "Show thousands separators" enabled? Happens in master too if you just enter a long enough number so that thousand separators are shown.
This is caused by call to 
`gtk_text_buffer_insert(GTK_TEXT_BUFFER(equation), &iter, buffer, -1);`
in function
`static void reformat_separators(MathEquation *equation)`
in file math-equation.c.

**Comment by sc0w** [comment]
@mbkma yes, you are right

**Comment by raveit65** [comment]
@mbkma 
open question in previous post.

**Comment by mbkma** [comment]
@lukefromdc yes, you are right. Thanks for pointing this out. final_answer has to be freed, as the return type of `mp_cast_to_string` is a gchar* which was allocated using `g_strndup`. I also replaced the error message with a more specific one.

**Comment by lukefromdc** [comment]
Just realized: since this is a new feature we need to bump the dev version to 1.25 and keep it out of version 1.24 point releases unless others want to add this for 1.24.1

**Pull request merged**